Draft release notes for Mimir 2.9.0 (#5122)
grafana/mimir: docs/sources/mimir/release-notes/v2.9.md

---
title: "Grafana Mimir version 2.9 release notes"
menuTitle: "V2.9 release notes"
description: "Release notes for Grafana Mimir version 2.9"
weight: 300
---

# Grafana Mimir version 2.9 release notes

Grafana Labs is excited to announce version 2.9 of Grafana Mimir.

The highlights that follow include the top features, enhancements, and bugfixes in this release. For the complete list of changes, see the [changelog](https://github.com/grafana/mimir/blob/main/CHANGELOG.md).

## Features and enhancements

> - **Experimental support for optimized series selection strategy** Mimir can now use an optimized strategy for fetching posting lists that can significantly reduce memory allocations in the store-gateway.

Reviewer: This is a great optimization but this description feels a little dense for release notes. Maybe @dimitarvdimitrov can distill it down?

Reviewer 2: ```suggestion
- **Experimental support for optimized series selection strategy** The store-gateway can now minimize the data necessary to complete requests. Based on the query it can fetch slightly more series and filter them down them before returning as opposed to spending resources to narrow down the selection beforehand.
```

How about this? Alternatively we can just say that the store-gateway can now complete requests with fetching less data.
```suggestion
- **Experimental support for optimized series selection strategy** Reduce the volume of data necessary to complete requests with individual label selectors that are present on a lot of series.
```

I think I prefer the latter.

- **Reduced CPU utilisation for some shuffle sharding scenarios** Mimir queriers will now use significantly less CPU in cases where shuffle sharding is enabled with a high number of shards.
- **Reduced object storage API calls in compactors and rulers** Mimir 2.9 comes with optimizations that will reduce the amount of times compactors and rulers need to access rules stored in object storage.

> - **Experimental support for streaming chunks from ingester to querier** This is expected to greatly reduce querier memory consumption when evaluating queries that select a large number of chunks, because chunks streamed from the querier can now be read into memory as needed.

Reviewer 3: This will only be beneficial for queries that select many series (we still load all chunks for a single series into memory at once)

```suggestion
- **Experimental support for streaming chunks from ingester to querier** This is expected to greatly reduce querier memory consumption when evaluating queries that select a large number of series, because chunks streamed from the querier can now be read into memory as needed.
```

### Helm chart improvements

The Grafana Mimir and Grafana Enterprise Metrics Helm chart is now released independently. See the [Grafana Mimir Helm chart documentation](/docs/helm-charts/mimir-distributed/latest/).

## Important changes

In Grafana Mimir 2.9 we have removed the following previously deprecated or experimental metrics:

- `cortex_bucket_store_chunk_pool_requested_bytes_total`
- `cortex_bucket_store_chunk_pool_returned_bytes_total`

> The following configuration options are deprecated and will be removed in Grafana Mimir 2.11:

Reviewer 3: There will be two more to add to this list (`-querier.iterators` and `-query.batch-iterators`) after https://github.com/grafana/mimir/pull/5114 is merged

---
title: "Grafana Mimir version 2.9 release notes"
menuTitle: "V2.9 release notes"
description: "Release notes for Grafana Mimir version 2.9"
weight: 300
---

# Grafana Mimir version 2.9 release notes

Grafana Labs is excited to announce version 2.9 of Grafana Mimir.

The highlights that follow include the top features, enhancements, and bugfixes in this release. For the complete list of changes, see the [changelog](https://github.com/grafana/mimir/blob/main/CHANGELOG.md).

## Features and enhancements

> - **Experimental support for optimized series selection strategy** Reduce the volume of data necessary to complete requests with individual label selectors that are present on a lot of series.

Reviewer 4: I think this is not much clear. I would clarify:
1. This applies to metrics fetched from the long-term storage via the store-gateway
2. Expand what means "individual label selectors that are present on a lot of series"

Reviewer 2: How about this? I think an example is easier to grasp than explaining exactly what this means.

```suggestion
- **Reduced store-gateway memory utilization on fetching series from long-term storage** Reduce the volume of data necessary to evaluate queries which select series by labels which are present on many series, such as `datacenter="nyc"`.
```

Author: How about
```suggestion
- **Experimental support for optimized series selection strategy** For queries with broad label matchers (e.g. `datacenter="dc1"), Mimir 2.9 will fetch a reduced amount of index data, which leads to a significant reduction in memory allocations in the store-gateway.
```

> - **Reduced CPU utilisation for some shuffle sharding scenarios** Mimir queriers will now use significantly less CPU in cases where shuffle sharding is enabled with a high number of shards.

Reviewer 4: ```suggestion
- **Reduced CPU utilisation for some shuffle sharding scenarios** Mimir queriers will now use significantly less CPU in cases where shuffle sharding is enabled for tenants with a large shard size but lower than the total number of ingesters.
```

> - **Reduced object storage API calls in compactors and rulers** Mimir 2.9 comes with optimizations that will reduce the amount of times compactors and rulers need to access rules stored in object storage.

Reviewer 4: We need to expand here. I would include:
- Added experimental ruler storage cache support. The cache should reduce the number of "list objects" API calls issued to the object storage when there are 2+ ruler replicas running in a Mimir cluster. The cache can be configured setting `-ruler-storage.cache.*` CLI flags or their respective YAML config options.
- Introduced a new feature to trigger a synchronization of tenant's rule groups as soon as they change the rules configuration via API. This synchronization is in addition of the periodic syncing done every `-ruler.poll-interval`. The new behavior is enabled by default, but can be disabled with `-ruler.sync-rules-on-changes-enabled=false` (configurable on a per-tenant basis too).

Author: I added two bullet points for the individual enhancements.

> - **Experimental support for streaming chunks from ingester to querier** This is expected to greatly reduce querier memory consumption when evaluating queries that select a large number of series, because chunks streamed from the querier can now be read into memory as needed.

Reviewer 4: @charleskorn Do you want to add more color here?

Reviewer 3: I don't have any more specifics to add just yet - once we've done some more testing we might have something more to add. 

I'm OK for this to go as-is, we can talk about it more in the next release (and the next release will also include a bunch of improvements that didn't make this release).

## Features and enhancements

- **Reduced store-gateway memory utilization on fetching series from long-term storage** For queries that include broad label matchers (e.g. `datacenter="dc1"`), Mimir 2.9 will fetch a reduced volume of index data, which leads to a significant reduction in memory allocations in the store-gateway.
- **Reduced CPU utilisation for some shuffle sharding scenarios** Mimir queriers will now use significantly less CPU in cases where shuffle sharding is enabled for tenants with a shard size that's large but lower than the total number of ingesters.
- **Reduced object storage API calls in compactors and rulers** Mimir 2.9 comes with optimizations that will reduce the amount of times compactors and rulers need to access rules stored in object storage.
  - This release adds experimental support for a ruler storage cache. This cache should reduce the number of "list objects" API calls issued to the object storage when there are 2+ ruler replicas running in a Mimir cluster. The cache can be configured by setting the `-ruler-storage.cache.*` CLI flags or their respective YAML config options.

>   - We also introduced a new feature to trigger a synchronization of tenant's rule groups as soon as changes to the rule configuration are made via API. This synchronization is in addition of the periodic syncing done every -ruler.poll-interval and allows to increase the polling interval. The new behavior is enabled globally by default but can be disabled with `-ruler.sync-rules-on-changes-enabled=false` or tuned at a per-tenant level.

Reviewer 4: ```suggestion
  - We also introduced a new feature to trigger a synchronization of tenant's rule groups as soon as changes to the rule configuration are made via API. This synchronization is in addition of the periodic syncing done every `-ruler.poll-interval`, which has then be relaxed from every `1m` to every `10m`. The new behaviour is enabled globally by default but can be disabled with `-ruler.sync-rules-on-changes-enabled=false` or tuned at a per-tenant level.
```
